src/ibex_bluesky_core/plans/reflectometry/autoalign_utils.py

```diff
@@ -0,0 +1,357 @@
+
+    Args:
+        name (str): The name of the alignment parameter.
+            Should be same as the respective relectometry parameter or read/write block.
+        rel_scan_ranges (list[float]): Scan range relative to the current motor position.
+            If the list has more than one element then it will rescan for each range.
+            Scans between mot_pos - rel_scan_range / 2 -> mot_pos + rel_scan_range / 2
+        fit_method (FitMethod): The relationship to expect between the alignment parameter
+            and the beam. e.g Gaussian.
+        fit_param (str): Which property of fit_method you aim to optimise. e.g centre (x0)
+            of a Gaussian. See fitting/fitting_utils for the possible options for each fitting
+            method.
+        pre_align_param_positions (dict[str, float], optional): A dictionary of alignment
+            parameter names to values. Before alignment, each supplied alignment parameter
+            will be moved to their respective value.
+        check_func (Callable[[ModelResult, float], bool], optional): Checks to ensure that the
+            optimised value for this alignment parameter is sensible, must return True if NOT
+            a sensible value.
+        _prefix (str, optional): The PV prefix for the respective reflectometry parameter.
+            Defaults to the current instrument PV prefix.
+        _movable (ReflParameter, optional): A reflectometry parameter or block
+            read/write signal. Use this if you want to create your own signal.
+
+    """
+
+    name: str
+    rel_scan_ranges: list[float]
+    fit_method: FitMethod
+    fit_param: str
+    pre_align_param_positions: dict[str, float] = field(default_factory=dict)
+    check_func: Callable[[ModelResult, float], bool] | None = None
+    _prefix: str | None = None
+    _movable: ReflParameter | None = None
+
+    def get_movable(self) -> ReflParameter:
+        """Return the encapsulated signal for this alignment parameter.
+
+        If there is no signal attached then one is created as a reflectometry parameter.
+
+        Returns:
+            signal (ReflParameter)
+
+        """
+        if not self._movable:
+            self._movable = ReflParameter(
+                prefix=self._prefix or get_pv_prefix(), name=self.name, changing_timeout_s=60
+            )
+
+        return self._movable
+
+    def pre_align_params(
+        self,
+        alignment_params: list["AlignmentParam"],
+    ) -> Generator[Msg, None, None]:
+        """Move the alignment parameters to respective values in pre_align_param_positions.
+
+        Args:
+            alignment_params (list[AlignmentParam]): The alignment parameters to be moved.
+
+        """
+        yield from bps.mv(
+            *chain(
+                *[
+                    (p.get_movable(), self.pre_align_param_positions[p.name])
+                    for p in alignment_params
+                ]
+            )  # type: ignore
+        )
+
+
+def _add_fields(
```

Comment:
this part is sort of duplicated by https://github.com/IsisComputingGroup/ibex_bluesky_core/blob/c85e1a91060c5af252e5d5aec4b2aa2c73cb8565/src/ibex_bluesky_core/plans/__init__.py#L70 but maybe we should pull the fields bit out so you can just pass in more stuff (ie `alignment_param_name`) - up to you really